libnavigation-base/src/main/java/com/mapbox/navigation/base/utils/DecodeUtils.kt

```diff
@@ -0,0 +1,108 @@
+package com.mapbox.navigation.base.utils
+
+import com.mapbox.api.directions.v5.DirectionsCriteria
+import com.mapbox.api.directions.v5.models.DirectionsRoute
+import com.mapbox.api.directions.v5.models.LegStep
+import com.mapbox.core.constants.Constants
+import com.mapbox.geojson.LineString
+import com.mapbox.geojson.Point
+import com.mapbox.geojson.utils.PolylineUtils
+import com.mapbox.navigation.base.internal.utils.CacheResultUtils.cacheResult
+
+object DecodeUtils {
+
+    private val completeGeometryDecodeCache = PolylineUtils::decode.cacheResult(maxSize = 3)
+    private val stepsGeometryDecodeCache = PolylineUtils::decode.cacheResult(maxSize = 50)
```

Comment:
This needs to be much bigger. Trying now a route from Poland to the south of Spain I can see more than 130 steps. We could also have up to 2 alternatives these days (and more in the future). When I tried excluding highways and toll roads, this grows to nearly 1000 steps per route.

This number seems to be very much scenario dependent and it might not make sense to set up a cache for 5k elements to handle an edge case while normally we'd see the usage of dozens of elements at most (for local and short routes).

Should we take another look at trying to limit the number of cached elements per number of routes?

Reply:
I change the functions, so that the user is now required to pass a route along with a leg step. 
This way we can know which routes are currently used for decoding steps' geometries and are able to resize the cache accordingly.